elastic/elastic-agent
# Fix: Agent failed to upgrade from 8.4.2 to 8.5.0 BC1 for MAC 12 agent using agent binary.

## What does this PR do?

* Reverts the previous revert commit, restoring the elastic-agent app bundle for Mac OS.
* Implements the workaround for upgrade from pre 8.5 version to 8.5. (assuming that this fix can go into 8.5)

This workaround is ONLY activated when upgrading from pre 8.5 version to 8.5+ on for Mac OS distribution.
The workaround is not used or affects the clean install, nor the it's not going used for any upgrades after 8.5.

For upgrades after 8.5 the symlink will be correctly established from the upgrade handle:
<img width="2112" alt="Screen Shot 2022-10-01 at 2 37 20 PM" src="https://user-images.githubusercontent.com/872351/193464596-d57aaad3-f81a-4b9b-baf3-59bce4e9eb5a.png">


Here is how is how the workaround works for let say the upgrade from 8.4 to 8.5.
1. The 8.5 package is bundled with the additional executable shell script that is unpacked into what was previously the  ```elastic-agent``` binary location. The sole purpose of the shell script is to fix the symlink created by pre 8.5 versions of upgrade handle.
<img width="793" alt="Screen Shot 2022-10-02 at 12 07 28 PM" src="https://user-images.githubusercontent.com/872351/193464158-1e842b3f-0983-4d8a-a749-9fb5e728af6b.png">

2. When the upgrade from 8.4 is initiated the 8.5 is unpacked, and the 8.4 upgrade handle creates the symlink to this new ```elastic-agent``` shell script.
3. The 8.5 elastic agent is started, the launchd executes the ```elastic-agent``` shell script that fixes up the symlink to point correctly to the binary inside of the elastic agent app bundle and exits.
4. The launchd restarts the 8.5 if it exists (KeepAlive is set to true) and this time the symlink is correctly pointed to the binary inside of the bundle.
5. The actual ```elastic-agent``` comes up and the upgrade completes.

I tested the upgrades on all 3 platforms macOS and as a regression test on linux and windows:
1.  from 8.4 to 8.5 
2. from 8.5. to 8.6.

**This would need to be backported to main branch as well.**

## Why is it important?

Addresses the issue: "Agent failed to upgrade from 8.4.2 to 8.5.0 BC1 for MAC 12 agent using agent binary."
https://github.com/elastic/elastic-agent/issues/1298

## Checklist


## How to test this PR locally

I tested these changes with the local builds that had the artifactory download/verify code disabled so I could drop the newer locally built elastic-agent binaries into the downloads folder.

## Related issues

- Closes https://github.com/elastic/elastic-agent/issues/1298

## Screenshots

MacOS upgrade from 8.4 to 8.5

<img width="1303" alt="Screen Shot 2022-10-01 at 8 53 43 PM" src="https://user-images.githubusercontent.com/872351/193464929-d64a9f74-9503-4dd1-a1d7-1be1d520fdb0.png">
<img width="1111" alt="Screen Shot 2022-10-01 at 8 56 41 PM" src="https://user-images.githubusercontent.com/872351/193464936-97d6d37e-9ed1-4602-be33-8e2d04be8caf.png">
<img width="1301" alt="Screen Shot 2022-10-01 at 8 57 55 PM" src="https://user-images.githubusercontent.com/872351/193464941-ecdeab5d-5009-4362-8d44-1926c1da8b97.png">

MacOS upgrade from 8.5 to 8.6
<img width="1295" alt="Screen Shot 2022-10-01 at 9 12 55 PM" src="https://user-images.githubusercontent.com/872351/193464985-6a5ab8e2-d451-466b-8f9e-521aa1ab75e4.png">
<img width="1301" alt="Screen Shot 2022-10-01 at 9 27 52 PM" src="https://user-images.githubusercontent.com/872351/193464990-79c1e2d9-516c-464f-8826-294679d1b47c.png">



Linux (Ubuntu) upgrade from 8.4 to 8.5, regression test
<img width="1297" alt="Screen Shot 2022-10-02 at 9 44 10 AM" src="https://user-images.githubusercontent.com/872351/193464999-a8bc3a13-78eb-42b1-9a90-34382642a24d.png">
<img width="1274" alt="Screen Shot 2022-10-02 at 9 45 10 AM" src="https://user-images.githubusercontent.com/872351/193465006-2e10d937-5977-4d0a-9698-942d79771631.png">

Linux (Ubuntu) upgrade from 8.5 to 8.6, regression test
<img width="1307" alt="Screen Shot 2022-10-02 at 10 26 08 AM" src="https://user-images.githubusercontent.com/872351/193465038-fcc55b30-d429-4cc2-b599-10867a5c659c.png">
<img width="1300" alt="Screen Shot 2022-10-02 at 10 27 16 AM" src="https://user-images.githubusercontent.com/872351/193465042-b61b5ae6-9b43-4dd7-970f-772c44c2712c.png">

Windows (10) upgrade from 8.4 to 8.5, regression test
<img width="1299" alt="Screen Shot 2022-10-02 at 11 31 17 AM" src="https://user-images.githubusercontent.com/872351/193465066-b038b2a3-eaf6-4167-8a68-a570cdad7a0c.png">
<img width="1297" alt="Screen Shot 2022-10-02 at 11 32 10 AM" src="https://user-images.githubusercontent.com/872351/193465075-7219c9d3-c0a9-42a4-bb36-eaedadcb2db9.png">

Windows (10) upgrade from 8.5 to 8.6, regression test
<img width="1308" alt="Screen Shot 2022-10-02 at 11 44 45 AM" src="https://user-images.githubusercontent.com/872351/193465100-f13cd03a-4bee-49f1-b0a0-47d107a112c9.png">
<img width="1298" alt="Screen Shot 2022-10-02 at 11 45 37 AM" src="https://user-images.githubusercontent.com/872351/193465105-d43ea55b-78f5-4c1a-bbda-31e3256e25cf.png">
